lemminx-liberty/src/main/java/io/openliberty/tools/langserver/lemminx/services/FeatureService.java

```diff
@@ -374,21 +377,40 @@ public List<Feature> readFeaturesFromFeatureListFile(List<Feature> installedFeat
         
         // Note: Only the public features are loaded when unmarshalling the passed featureListFile.
         if ((featureInfo.getFeatures() != null) && (featureInfo.getFeatures().size() > 0)) {
-            for (int i = 0; i < featureInfo.getFeatures().size(); i++) {
-                Feature f = featureInfo.getFeatures().get(i);
+            FeatureListGraph featureListGraph = new FeatureListGraph();
+            for (Feature f : featureInfo.getFeatures()) {
                 f.setShortDescription(f.getDescription());
                 // The xml featureListFile does not have a wlpInformation element like the json does, but our code depends on looking up 
                 // features by the shortName found in wlpInformation. So create a WlpInformation object and initialize the shortName to 
                 // the feature name.
                 WlpInformation wlpInfo = new WlpInformation(f.getName());
                 f.setWlpInformation(wlpInfo);
+
+                String currentFeature = f.getName();            
+                List<String> enables = f.getEnables();
+                List<String> configElements = f.getConfigElements();
+                FeatureListNode currentFeatureNode = featureListGraph.addFeature(currentFeature);
+                if (enables != null) {
+                    for (String enabledFeature : enables) {
+                        FeatureListNode feature = featureListGraph.addFeature(enabledFeature);
+                        feature.addEnabledBy(currentFeature);
+                        currentFeatureNode.addEnables(enabledFeature);
+                    }
+                }
+                if (configElements != null) {
+                    for (String configElement : configElements) {
+                        FeatureListNode configNode = featureListGraph.addConfigElement(configElement);
+                        configNode.addEnabledBy(currentFeature);
+                        currentFeatureNode.addEnables(configElement);
+                    }
+                }
             }
             installedFeatures = featureInfo.getFeatures();
+            libertyWorkspace.setFeatureListGraph(featureListGraph);
             libertyWorkspace.setInstalledFeatureList(installedFeatures);
         } else {
             LOGGER.warning("Unable to get installed features for current Liberty workspace: " + libertyWorkspace.getWorkspaceString());
```

Comment:
Should the featureListGraph in the libertyWorkspace get cleared out in this case?